x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/token/TransportCreateTokenAction.java

```diff
@@ -64,26 +70,59 @@ protected void doExecute(Task task, CreateTokenRequest request, ActionListener<C
         }
     }
 
-    private void authenticateAndCreateToken(CreateTokenRequest request, ActionListener<CreateTokenResponse> listener) {
+    private void authenticateAndCreateToken(GrantType grantType, CreateTokenRequest request, ActionListener<CreateTokenResponse> listener) {
         Authentication originatingAuthentication = Authentication.getAuthentication(threadPool.getThreadContext());
         try (ThreadContext.StoredContext ignore = threadPool.getThreadContext().stashContext()) {
-            final UsernamePasswordToken authToken = new UsernamePasswordToken(request.getUsername(), request.getPassword());
+            final AuthenticationToken authToken = extractAuthenticationToken(grantType, request, listener);
+            if (authToken == null) {
+                listener.onFailure(new IllegalStateException(
+                        "grant_type [" + request.getGrantType() + "] is not supported by the create token action"));
+                return;
+            }
+
             authenticationService.authenticate(CreateTokenAction.NAME, request, authToken,
                 ActionListener.wrap(authentication -> {
-                    request.getPassword().close();
+                    clearCredentialsFromRequest(grantType, request);
+
                     if (authentication != null) {
```

Comment:
Can you please elaborate @bizybot for additional context ? What is this token ? What would the client do with it ?

> extract the thread context response header and pass it inside create token response body
The header value from ElasticsearchSecurityException needs to be extracted in RestGetTokenAction and create oauth2 error response, but there is a problem as there is no appropriate error code to use.

I don't mind us defining a new error_code, since we are defining a new grant type

> or pass this as response header WWW-Authenticate as is which the Kibana needs to intercept and forward it to the browser client. This is the same for successful or unsuccessful authentication. 

Wouldn't we need to parse the Exception either way here too in order to populate the `WWW-Authenticate` response header?

I'd wait for some more details but with I can understand now, since we might need to return this token in cases of both successful and failed authentication attempts, it wouldn't seem natural to return this as `error`